cartographer/io/probability_grid_points_processor.cc

```diff
@@ -0,0 +1,147 @@
+
+  for (int y = 0; y < mat.rows(); ++y) {
+    for (int x = 0; x < mat.cols(); ++x) {
+      const PixelData& pixel = mat(y, x);
+      const int r = pixel.r;
+      const int g = pixel.g;
+      const int b = pixel.b;
+      image.pixels[y * image.stride / 4 + x] = (255 << 24) | (r << 16) | (g << 8) | b;
+    }
+  }
+
+  WritePng(&image, file_writer);
+  CHECK(file_writer->Close());
+}
+
+mapping_2d::ProbabilityGrid CreateProbabilityGrid(double resolution) {
+  constexpr int kInitialProbabilityGridSize = 100;
+  Eigen::Vector2d max = 0.5 * kInitialProbabilityGridSize
+    * resolution * Eigen::Vector2d::Ones();
+  return mapping_2d::ProbabilityGrid(
+    cartographer::mapping_2d::MapLimits(
+      resolution, max,
+      mapping_2d::CellLimits(kInitialProbabilityGridSize,
+                             kInitialProbabilityGridSize)));
+}
+
+}  // namespace
+
+ProbabilityGridPointsProcessor::ProbabilityGridPointsProcessor(
+  const double resolution,
+    const mapping_2d::proto::RangeDataInserterOptions&
+    range_data_inserter_options,
+  std::unique_ptr<FileWriter> file_writer,
+  PointsProcessor* const next)
+  : next_(next),
+    file_writer_(std::move(file_writer)),
+    range_data_inserter_(range_data_inserter_options),
+    probability_grid_(CreateProbabilityGrid(resolution)) {}
+
+std::unique_ptr<ProbabilityGridPointsProcessor>
+ProbabilityGridPointsProcessor::FromDictionary(
+  FileWriterFactory file_writer_factory,
+  common::LuaParameterDictionary* const dictionary,
+  PointsProcessor* const next) {
+  return common::make_unique<ProbabilityGridPointsProcessor>(
+    dictionary->GetDouble("resolution"),
+      mapping_2d::CreateRangeDataInserterOptions(
+          dictionary->GetDictionary("range_data_inserter").get()),
+    file_writer_factory(dictionary->GetString("filename") + ".png"), next);
+}
+
+void WriteGrid(const mapping_2d::ProbabilityGrid & probability_grid,
+               FileWriter* const file_writer) {
+  Eigen::Array2i offset;
+  mapping_2d::CellLimits cell_limits;
+  probability_grid.ComputeCroppedLimits(&offset, &cell_limits);
+  if (cell_limits.num_x_cells == 0 || cell_limits.num_y_cells == 0)
+  {
+    LOG(WARNING) << "Not writing output: empty probability grid";
+    return;
+  }
+
+  const auto grid_index_to_pixel = [cell_limits](const Eigen::Array2i& index) {
+    return Eigen::Array2i(cell_limits.num_x_cells - index(0) - 1,
+                          cell_limits.num_y_cells - index(1) - 1);
+  };
+  int width = cell_limits.num_y_cells;
+  int height = cell_limits.num_x_cells;
+  PixelDataMatrix image(height, width);
+  for (auto xy_index :
+         cartographer::mapping_2d::XYIndexRangeIterator(cell_limits)) {
+    auto index = xy_index + offset;
+    if (probability_grid.IsKnown(index)) {
+      const Eigen::Array2i pixel = grid_index_to_pixel(xy_index);
+      PixelData& pixel_data = image(pixel.x(), pixel.y());
+      float probability = probability_grid.GetProbability(index);
```

Comment:
can be const. Also to stay consistent, this should be `1.f - probability_grid...`.